File: selection_sort/python/selection.py
import sys
import time
import os

def selection_sort(arr):
    n = len(arr)
    for i in range(n):
        min_idx = i
        for j in range(i + 1, n):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]

def rodar_teste(nome_arquivo, n):
    try:
        with open(nome_arquivo, 'r') as f:
            v = []
            for _ in range(n):
                line = f.readline()
                if not line:
                    break
                v.append(float(line.strip()))

        if len(v) < n:
            print(f"Aviso: O arquivo {nome_arquivo} possui apenas {len(v)} registros.")

        start = time.perf_counter()
        selection_sort(v)
        end = time.perf_counter()

        return end - start
    except FileNotFoundError:
        print(f"Erro: Arquivo {nome_arquivo} não encontrado.")
        return -1
    except Exception as e:
        print(f"Erro: {e}")
        return -1

def main():
    if len(sys.argv) < 2:
        print(f"Uso: python3 {sys.argv[0]} <n_elementos>")
        sys.exit(1)

    n = int(sys.argv[1])
    base_path = os.path.dirname(__file__)

    print(f"--- Iniciando Benchmark Python (N: {n}) ---\n")

    testes = [
        ("Crescente",   os.path.join(base_path, "..", "dados", "crescente.txt"),   "(Melhor caso)"),
        ("Aleatorio",   os.path.join(base_path, "..", "dados", "aleatorio.txt"),   "(Caso Médio)"),
        ("Decrescente", os.path.join(base_path, "..", "dados", "decrescente.txt"), "(Pior caso)"),
    ]

    for i, (nome, path, desc) in enumerate(testes):
        tempo = rodar_teste(path, n)
        if tempo >= 0:
            print(f"{i + 1}. {nome:12}: {tempo:.6f} segundos {desc}")

    print("\n--- Testes concluidos ---")

if __name__ == "__main__":
    main()

File: selection_sort/javascript/selection.js
const fs = require('fs');
const path = require('path');
const { performance } = require('perf_hooks');

function selectionSort(arr) {
    let n = arr.length;
    for (let i = 0; i < n; i++) {
        let minIdx = i;
        for (let j = i + 1; j < n; j++) {
            if (arr[j] < arr[minIdx]) {
                minIdx = j;
            }
        }
        // Troca (Swap)
        [arr[i], arr[minIdx]] = [arr[minIdx], arr[i]];
    }
}

/**
 * Função para carregar dados e medir o tempo.
 */
function rodarTeste(nomeArquivo, n) {
    try {
        const data = fs.readFileSync(nomeArquivo, 'utf8');
        let v = data.split('\n')
                    .filter(line => line.trim() !== '')
                    .slice(0, n)
                    .map(Number);

        if (v.length < n) {
            console.log(`Aviso: O arquivo ${nomeArquivo} possui apenas ${v.length} registros.`);
        }

        const start = performance.now();
        selectionSort(v);
        const end = performance.now();

        // performance.now() retorna milissegundos, convertemos para segundos
        return (end - start) / 1000;
    } catch (err) {
        console.error(`Erro ao abrir ${nomeArquivo}:`, err.message);
        return -1;
    }
}

function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.log("Uso: node selection.js <n_elementos>");
        process.exit(1);
    }

    const n = parseInt(args[0]);
    console.log(`--- Iniciando Benchmark JavaScript (N: ${n}) ---\n`);

    const base = path.join(__dirname, '..', 'dados');

    const testes = [
        { nome: "Crescente",   path: path.join(base, "crescente.txt"),   desc: "(Melhor caso)" },
        { nome: "Aleatorio",   path: path.join(base, "aleatorio.txt"),   desc: "(Caso Médio)"  },
        { nome: "Decrescente", path: path.join(base, "decrescente.txt"), desc: "(Pior caso)"   },
    ];

    testes.forEach((teste, index) => {
        const tempo = rodarTeste(teste.path, n);
        if (tempo >= 0) {
            console.log(`${index + 1}. ${teste.nome.padEnd(12)}: ${tempo.toFixed(6)} segundos ${teste.desc}`);
        }
    });

    console.log("\n--- Testes concluidos ---");
}

main();